lib/stages/bufferDelay.cpp

```diff
@@ -0,0 +1,104 @@
+#include "bufferDelay.hpp"
+
+#include "StageFactory.hpp"   // for REGISTER_KOTEKAN_STAGE, StageMakerTemplate
+#include "kotekanLogging.hpp" // for INFO, DEBUG2, FATAL_ERROR
+#include "visUtil.hpp"
+
+#include "fmt.hpp" // for format, fmt
+#include "json.hpp"
+
+#include <algorithm>  // for max
+#include <atomic>     // for atomic_bool
+#include <exception>  // for exception
+#include <functional> // for _Bind_helper<>::type, bind, function
+#include <regex>      // for match_results<>::_Base_type
+#include <stdexcept>  // for runtime_error, invalid_argument
+#include <stdint.h>   // for uint8_t
+#include <string.h>   // for memcpy
+
+
+using nlohmann::json;
+
+using kotekan::bufferContainer;
+using kotekan::Config;
+using kotekan::Stage;
+
+REGISTER_KOTEKAN_STAGE(bufferDelay);
+
+bufferDelay::bufferDelay(Config& config, const std::string& unique_name,
+                       bufferContainer& buffer_container) :
+    Stage(config, unique_name, buffer_container, std::bind(&bufferDelay::main_thread, this)) {
+
+    in_buf = get_buffer("in_buf");
+    register_consumer(in_buf, unique_name.c_str());
+
+    out_buf = get_buffer("out_buf");
+    register_producer(out_buf, unique_name.c_str());
+    
+    _copy_frame = config.get_default<bool>(unique_name, "copy_frame", false);
+    _num_frames_to_hold = config.get<uint32_t>(unique_name, "num_frames_to_hold");
+
+    if (_num_frames_to_hold > (uint32_t)in_buf->num_frames) {
+      throw std::invalid_argument(fmt::format(fmt("The no. of frames to hold ({:d}) is greater than the number of frames that can be stored in the input buffer ({:d})."),
+            _num_frames_to_hold, in_buf->num_frames));
+    }
+
+}
+
+void bufferDelay::main_thread() {
+
+    //frameID in_frame_id(in_buf);
+    uint32_t in_frame_id = 0;
+    uint32_t in_frame_hold_ctr = 0;
+    uint32_t in_frame_release_id = 0;
+    uint32_t out_frame_id = 0;
+
+    uint8_t* output_frame =
+      wait_for_empty_frame(out_buf, unique_name.c_str(), out_frame_id);
+    if (output_frame == nullptr)
+      return; // Shutdown condition
```

Comment:
I don't see why you need to get an output frame here.  You can just get it as needed in the loop below as needed.